functions.php

```diff
@@ -24,6 +24,23 @@ function wpm_apply_filters_typesafe( $hook_name, $value, ...$args ) {
 function wpm_apply_filters_typed( $type, $hook_name, $value, ...$args ) {
 	$next_value = apply_filters( $hook_name, $value, ...$args ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.DynamicHooknameFound
 
+	// Union types are separated by a pipe character.
+	if ( preg_match( '/\|/i', $type ) ) {
+		$types = explode( '|', $type );
+
+		foreach ( $types as $type ) {
+			if ( wpm_is_type( $type, $next_value ) ) {
+				return $next_value;
+			}
+		}
+
+		// None of the types matched.
+		_doing_it_wrong( __FUNCTION__, sprintf( 'Return value of "%1$s" filter must be of the type "%2$s", "%3$s" returned.', esc_attr( $hook_name ), esc_attr( $type ), esc_attr( gettype( $next_value ) ) ), '1.0.1' );
+
+		return $value;
+	}
+
+	// Single type.
 	if ( ! wpm_is_type( $type, $next_value ) ) {
 		_doing_it_wrong( __FUNCTION__, sprintf( 'Return value of "%1$s" filter must be of the type "%2$s", "%3$s" returned.', esc_attr( $hook_name ), esc_attr( $type ), esc_attr( gettype( $next_value ) ) ), '1.0.1' );
```

Comment:
Maybe we could avoid having one part of the code for multiple types and one part for single types. WDYT of something like:

```
		// Union types are separated by a pipe character.
		$types = [$type]
		if ( preg_match( '/\|/i', $type ) ) {
    			$types = explode( '|', $type );
		}
		foreach ( $types as $type ) {
                 	if ( wpm_is_type( $type, $next_value ) ) {
				return $next_value;
			}
		}

		// None of the types matched.
		_doing_it_wrong( __FUNCTION__, sprintf( 'Return value of "%1$s" filter must be of the type "%2$s", "%3$s" returned.', esc_attr( $hook_name ), esc_attr( $type ), esc_attr( gettype( $next_value ) ) ), '1.0.1' );

		return $value;
```